# Assignment 5


Your task in this week’s assignment is to answer three questions:
1. What is the northernmost airport in the United States?
2. What is the easternmost airport in the United States?
3. On February 12th, 2013, which New York area airport had the windiest weather?
Being successful on this assignment requires using both:
• Your developing skills in Python and pandas, and
• Your research skills in validating your candidate answers in context.

### Data

In [2]:
# Import libraries
import pandas as pd
import sys

In [65]:
# airlines
files = ['airports', 'weather']

dfs = {}
for f in files:
    print(f)
    dfs[f] = pd.read_csv(f'https://raw.githubusercontent.com/hadley/nycflights13/master/data-raw/{f}.csv')
    
print(dfs.keys())

airports
weather
dict_keys(['airports', 'weather'])


# Analysis

## 1. What is the northernmost airport in the United States?

In [66]:
dfs["airports"].iloc[dfs["airports"]['lat'].idxmax()]

faa                          EEN
name     Dillant Hopkins Airport
lat                      72.2708
lon                      42.8983
alt                          149
tz                            -5
dst                            A
tzone                        NaN
Name: 417, dtype: object

In [67]:
dfs["airports"].sort_values(by=['lat'], ascending=False).head()

faa                                        name        lat         lon  \
417  EEN                     Dillant Hopkins Airport  72.270833   42.898333   
230  BRW                  Wiley Post Will Rogers Mem  71.285446 -156.766003   
110  AIN                          Wainwright Airport  70.638056 -159.994722   
708  K03                               Wainwright As  70.613378 -159.860350   
152  ATK  Atqasuk Edward Burnell Sr Memorial Airport  70.467300 -157.436000   

     alt  tz dst              tzone  
417  149  -5   A                NaN  
230   44  -9   A  America/Anchorage  
110   41  -9   A  America/Anchorage  
708   35  -9   A  America/Anchorage  
152   96  -9   A  America/Anchorage

If we look at the header data sorted by "lat", we can see that Dillant Hopkins Airport is the northernmost airport, but when we check this data, we find that it is incorrect (lat and long were changed).
Then, the northernmost airport in US is Wiley Post Will Rogers Mem (71.285446, -156.766003)

## 2. What is the easternmost airport in the United States?


In [68]:
dfs["airports"].iloc[dfs["airports"]['lon'].idxmin()]

faa                    ADK
name          Adak Airport
lat                 51.878
lon               -176.646
alt                     18
tz                      -9
dst                      A
tzone    America/Anchorage
Name: 95, dtype: object

In [69]:
dfs["airports"].sort_values(by=['lon'], ascending=True).head()

faa              name        lat         lon  alt  tz dst  \
95    ADK      Adak Airport  51.878000 -176.646000   18  -9   A   
112   AKB      Atka Airport  52.220278 -174.206389   56  -9   A   
522   GAM   Gambell Airport  63.766800 -171.733000   27  -9   A   
1281  SVA  Savoonga Airport  63.686400 -170.493000   53  -9   A   
1251  SNP    St Paul Island  57.167333 -170.220444   63  -9   A   

                  tzone  
95    America/Anchorage  
112   America/Anchorage  
522   America/Anchorage  
1281  America/Anchorage  
1251  America/Anchorage

The easternmost airport in US is Adak Airport (51.878000, -176.646000)

## 3. On February 12th, 2013, which New York area airport had the windiest weather?

I will analyze the data to find how to relate it, and fiew York area

In [71]:
dfs["airports"].head()

faa                           name        lat        lon   alt  tz dst  \
0  04G              Lansdowne Airport  41.130472 -80.619583  1044  -5   A   
1  06A  Moton Field Municipal Airport  32.460572 -85.680028   264  -6   A   
2  06C            Schaumburg Regional  41.989341 -88.101243   801  -6   A   
3  06N                Randall Airport  41.431912 -74.391561   523  -5   A   
4  09J          Jekyll Island Airport  31.074472 -81.427778    11  -5   A   

              tzone  
0  America/New_York  
1   America/Chicago  
2   America/Chicago  
3  America/New_York  
4  America/New_York

In [72]:
dfs["weather"].head()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
0    EWR  2013      1    1     1  39.02  26.06  59.37     270.0    10.35702   
1    EWR  2013      1    1     2  39.02  26.96  61.63     250.0     8.05546   
2    EWR  2013      1    1     3  39.02  28.04  64.43     240.0    11.50780   
3    EWR  2013      1    1     4  39.92  28.04  62.21     250.0    12.65858   
4    EWR  2013      1    1     5  39.02  28.04  64.43     260.0    12.65858   

   wind_gust  precip  pressure  visib             time_hour  
0        NaN     0.0    1012.0   10.0  2013-01-01T06:00:00Z  
1        NaN     0.0    1012.3   10.0  2013-01-01T07:00:00Z  
2        NaN     0.0    1012.5   10.0  2013-01-01T08:00:00Z  
3        NaN     0.0    1012.2   10.0  2013-01-01T09:00:00Z  
4        NaN     0.0    1011.9   10.0  2013-01-01T10:00:00Z

In [80]:
dfs["weather"].shape

(26115, 15)

Filter by date 2013-02-12

In [76]:
df_feb12 = dfs["weather"][dfs["weather"]["time_hour"].str.contains('2013-02-12')]
df_feb12

origin  year  month  day  hour   temp   dewp   humid  wind_dir  \
1001     EWR  2013      2   11    19  42.98  42.98  100.00       0.0   
1002     EWR  2013      2   11    20  42.08  42.08  100.00     300.0   
1003     EWR  2013      2   11    21  42.98  42.80  100.00     310.0   
1004     EWR  2013      2   11    22  44.06  42.80  100.00     290.0   
1005     EWR  2013      2   11    23  44.60  42.80   93.35     300.0   
...      ...   ...    ...  ...   ...    ...    ...     ...       ...   
18429    LGA  2013      2   12    14  44.06  23.00   43.02     300.0   
18430    LGA  2013      2   12    15  44.06  21.92   41.10     260.0   
18431    LGA  2013      2   12    16  42.98  24.08   46.91     300.0   
18432    LGA  2013      2   12    17  42.98  23.00   44.83     260.0   
18433    LGA  2013      2   12    18  41.00  21.92   46.23     260.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
1001      0.00000        NaN     0.0    1004.7    0.5  2013-02-12T00:00:00Z  
1002      3.45234        NaN     0.0    1004.7    0.5  2013-02-12T01:00:00Z  
1003      9.20624        NaN     0.0       NaN    1.5  2013-02-12T02:00:00Z  
1004      9.20624        NaN     0.0       NaN    9.0  2013-02-12T03:00:00Z  
1005      8.05546        NaN     0.0       NaN   10.0  2013-02-12T04:00:00Z  
...           ...        ...     ...       ...    ...                   ...  
18429    20.71404   25.31716     0.0    1011.5   10.0  2013-02-12T19:00:00Z  
18430    17.26170   20.71404     0.0    1011.3   10.0  2013-02-12T20:00:00Z  
18431     9.20624        NaN     0.0    1010.4   10.0  2013-02-12T21:00:00Z  
18432    14.96014   21.86482     0.0    1013.6   10.0  2013-02-12T22:00:00Z  
18433    17.26170   24.16638     0.0    1014.7   10.0  2013-02-12T23:00:00Z  

[72 rows x 15 columns]

In [85]:
df_feb12["origin"].unique(), df_feb12.shape

(array(['EWR', 'JFK', 'LGA'], dtype=object), (72, 15))

#### In this point I found that weather data is only from 'EWR', 'JFK' and 'LGA'. This airportrs are in NY area, then I shouldn't filter them

In [96]:
df_feb12_grouped = df_feb12.groupby("origin")

wind_data = {
    "wind_speed": df_feb12_grouped["wind_speed"].max(),
    "wind_gust": df_feb12_grouped["wind_gust"].max(), 
    "wind_dir": df_feb12_grouped["wind_dir"].max()
}


for k in wind_data.keys():
    print(f"{k}: {wind_data[k]}\n\n")


wind_speed: origin
EWR    1048.36058
JFK      20.71404
LGA      23.01560
Name: wind_speed, dtype: float64


wind_gust: origin
EWR    31.07106
JFK    27.61872
LGA    31.07106
Name: wind_gust, dtype: float64


wind_dir: origin
EWR    310.0
JFK    310.0
LGA    320.0
Name: wind_dir, dtype: float64




#### I will get only necesary columns and exclude outliers

In [109]:
from scipy import stats
import numpy as np


# I will take only 3 columns
df_feb12_reduced = df_feb12[["wind_speed", "wind_gust", "wind_dir"]] 

# Then, computing percentiles.
low = .05
high = .95
quant_df = df_feb12_reduced.quantile([low, high])
print(quant_df)

# Next filtering values based on computed percentiles
df_feb12_reduced = df_feb12_reduced.apply(lambda x: x[(x>quant_df.loc[low,x.name]) & (x < quant_df.loc[high,x.name])], axis=0)

df_feb12_reduced = pd.concat([df_feb12.loc[:,'origin'], df_feb12_reduced], axis=1)
df_feb12_reduced


      wind_speed  wind_gust  wind_dir
0.05    3.452340  20.311267     244.5
0.95   21.231891  31.071060     305.5


origin  wind_speed  wind_gust  wind_dir
1001     EWR         NaN        NaN       NaN
1002     EWR         NaN        NaN     300.0
1003     EWR     9.20624        NaN       NaN
1004     EWR     9.20624        NaN     290.0
1005     EWR     8.05546        NaN     300.0
...      ...         ...        ...       ...
18429    LGA    20.71404   25.31716     300.0
18430    LGA    17.26170   20.71404     260.0
18431    LGA     9.20624        NaN     300.0
18432    LGA    14.96014   21.86482     260.0
18433    LGA    17.26170   24.16638     260.0

[72 rows x 4 columns]

In [110]:
df_feb12_grouped = df_feb12_reduced.groupby("origin")

wind_data = {
    "wind_speed": df_feb12_grouped["wind_speed"].max(),
    "wind_gust": df_feb12_grouped["wind_gust"].max(), 
    "wind_dir": df_feb12_grouped["wind_dir"].max()
}


for k in wind_data.keys():
    print(f"{k}: {wind_data[k]}\n\n")

wind_speed: origin
EWR    20.71404
JFK    20.71404
LGA    20.71404
Name: wind_speed, dtype: float64


wind_gust: origin
EWR    26.46794
JFK    27.61872
LGA    28.76950
Name: wind_gust, dtype: float64


wind_dir: origin
EWR    300.0
JFK    300.0
LGA    300.0
Name: wind_dir, dtype: float64




In [120]:
columns = ["wind_speed", "wind_gust", "wind_dir"]
for airport in df_feb12_reduced["origin"].unique():
    print(airport)
    text = ''
    for c in columns:
        value = df_feb12_reduced[df_feb12_reduced["origin"]==airport][c].max()
        text += f"{c}: {value}\n"

    print(f"{text}\n\n")
    

EWR
wind_speed: 20.71404
wind_gust: 26.467940000000002
wind_dir: 300.0



JFK
wind_speed: 20.71404
wind_gust: 27.61872
wind_dir: 300.0



LGA
wind_speed: 20.71404
wind_gust: 28.7695
wind_dir: 300.0





#### Looking at information avobe, we can say that the the windiest weather airport on February 12th, 2013 was LGA

LGA (LaGuardia Airport)